CANTroller2/src/qpid.h

```diff
@@ -447,12 +436,153 @@ float QPID::GetOutputSum() { return outputSum; }
 float QPID::GetOutputMin() { return outMin; }  // Soren
 float QPID::GetOutputMax() { return outMax; }  // Soren
 float QPID::GetOutputRange() { return outMax - outMin; }  // Soren
+float QPID::GetCenter() { return center; }  // Soren
 uint8_t QPID::GetMode() { return static_cast<uint8_t>(mode); }
 uint8_t QPID::GetDirection() { return static_cast<uint8_t>(action); }
 uint8_t QPID::GetPmode() { return static_cast<uint8_t>(pmode); }
 uint8_t QPID::GetDmode() { return static_cast<uint8_t>(dmode); }
 uint8_t QPID::GetAwMode() { return static_cast<uint8_t>(iawmode); }
 uint8_t QPID::GetCentMode() { return static_cast<uint8_t>(centmode); }  // Soren
-float QPID::GetCenter() { return center; }  // Soren
 
+class IdleControl {  // Soren - To allow creative control of PID targets in case your engine idle problems need that.
+  public:
+    enum class idlemodes : uint32_t { direct, control, minimize, num_idlemodes };  // direct: disable idle management.  control: soft landing to idle rpm.  minimize: attempt to minimize idle to edge of instability
+  protected:
+    enum targetstates : uint32_t { driving, droptohigh, droptolow, idling, stallpoint };
+    // String modenames[3] = { "direct", "cntrol", "minimz" };
+    // String statenames[4] = { "drivng", "tohigh", "tolow", "tostal" };
+    targetstates runstate, laststate;
+    idlemodes idlemode;
+    float targetlast_rpm, idle_rpm, idlehigh_rpm, idlehot_rpm, idlecold_rpm, stallpoint_rpm, dynamic_rpm, temphot_f, tempcold_f;
+    float margin_rpm = 10; float idle_absmax_rpm = 1000.0;  // High limit of idle speed adjustability
+    float* target_rpm; float* measraw_rpm; float* measfilt_rpm; float* coolant_f;
+    bool we_just_changed_states = true;
+    uint32_t settletime_us;
+    Timer settleTimer;
+  public:
+    IdleControl (float* target_rpm_arg, float* measraw_rpm_arg, float* measfilt_rpm_arg, float* coolant_f_arg,  // Variable references: idle target, rpm raw, rpm filt, Engine temp
+      float idlehigh, float idlehot, float idlecold,  // Values for: high-idle rpm (will not stall), hot idle nominal rpm, cold idle nominal rpm 
+      float tempcold, float temphot,  // Values for: engine operational temp cold (min) and temp hot (max) in degrees-f
+      uint32_t settletime_us = 2000000,  // Period over which the idle will be lowered from high-idle to final idle
+      idlemodes idlemode = idlemodes::control) {  // Configure idle control to just soft land or also attempt to minimize idle
+        target_rpm = target_rpm_arg;
+        measraw_rpm = measraw_rpm_arg;
+        measfilt_rpm = measfilt_rpm_arg;
+        coolant_f = coolant_f_arg; 
+        set_idlehigh (idlehigh);
+        set_idlehot (idlehot);
+        stallpoint_rpm = idlehot_rpm;
+        set_idlecold (idlecold);
+        set_temphot (temphot);
+        set_tempcold (tempcold);        
+        calc_idlespeed();

```

Comment:
nit calc_idlespeed could be named idle_rpm or something to be clear whether it's talking about car speed or engine speed